### PR TITLE
Transformer-depth-scaling (ds-init) function

### PR DESCRIPTION
**Pull request by lkfo415579**
The depth-scaling is from [https://arxiv.org/abs/1908.11365](https://arxiv.org/abs/1908.11365). I implemented this one month ago, and i saw there is no implementation yet on the current commit. I have tested this function, it works.

### CODE REVIEW

**Comment by emjotde** [comment]
Gosh, I am sorry, but we have that already again :) `--transformer-depth-scaling` in master. 

And I actually came up with that a bit earlier than this paper in: https://arxiv.org/abs/1907.06170

**Comment by lkfo415579** [comment]
@emjotde Ohhhh, you actually are the one who first did this. but i didn't see any related code of transformer-depth-scaling option...

**Comment by emjotde** [comment]
The Edinburgh paper explains way better why it works, but we had that in some research branches as early as February or earlier. Let me check if the code actually does anything. I haven't ported everything from WMT19 research branch yet.

**Comment by lkfo415579** [comment]
@emjotde which branch is about WMT19 research, may i take a look as researching and reference? : >

**Comment by emjotde** [comment]
Ha! I put the option `--transformer-depth-scaling` in, but it's not doing anything in master. It works in `mjd/batch-scaling`. There is nothing "innovative" left in there, mostly fp16 training and changes around that. I am working now to have that moved over to master. The two branches have strongly diverged in the past half year. I plan to have them synced for version 1.20. There you should also see the depth-scaling related changes. 

Maybe let's sync before you add new features? I should be able to tell you what is "safe" to work on.

**Comment by lkfo415579** [comment]
@emjotde Sure, if the technique is quite the same, i should follow your instruction.

**Comment by emjotde** [comment]
The MATTN link is the link to my paper. About competence learning, we played around with that, but couldn't really reproduce the results for our larger data sets, did you have success there? If yes, then it's interesting.

**Comment by lkfo415579** [comment]
@emjotde sorry about the link. XD, it should be https://arxiv.org/abs/1908.11365 (merged attention sublayer (MAtt)) claims that in deep layers(20L) still can generate sentences as fast as base model.

Yeah, I tried wmt17 zh-en experiments with competence learning.
BLEU scores : Base[23.69], CL[24.30] on testset and 2.50x faster training time. Both of the results are using marian.

**Comment by emjotde** [comment]
If you already have it then we will be happy to help integrate it. I am not a fan of the AAN and prefer to use simplified RNNs in the decoder (better speed improvement than the AAN with e.g. an SSRU: https://www.aclweb.org/anthology/D19-5632/), but it's good to have.

**Comment by lkfo415579** [comment]
@emjotde I will take a look at SSRU first, thanks. If i had some spare time, i would provide CL learning code. XD

**Comment by emjotde** [comment]
`--transformer-decoder-autoreg rnn --dec-cell ssru` if you want to try it.

**Comment by lkfo415579** [comment]
@emjotde The paper is amazing, have you implemented Multi-Agent-Dual-Learning in Marian project? May i have a look?

**Comment by ykim362** [comment]
Thanks, @lkfo415579. MADL is more of a recipe than a functionality. All the necessary components are already in marian-dev. You can follow the steps in the paper. Please let us know if there's anything unclear.

**Comment by lkfo415579** [comment]
@ykim362 Accoring to the MADL paper, i see there are two models (f0 and g0) update gradient in the same time (**in alogirhtm 1, line 7**), the marian program has no any option about two models training together in one step, and the training objective function is a little different from CE. Therefore i am still confugsing about the correct training procedure of MADL, do you mind provide a **standard MADL marian training script** for me?

**Comment by ykim362** [comment]
@lkfo415579 I am sorry for the late reply. Even though our wngt work was inspired by MADL paper, we didn't follow the same algorithm as in the MADL paper. Rather, we used the method described in our wngt paper.

1. Train four forward direction teachers (different random seeds) with the original data
2. Train four backward direction teachers (different random seeds) with the original data
3. Generate 8-best translations of original source data with 4 forward teachers ensemble from step 1.
4. Generate 8-best translations of original target data with 4 backward teacher ensemble from step 2.
5. Pick the best translations among 8 candidates based on the sentence-level bleu score.

With these steps, you might have three different combinations of datasets. (3x amount)
1. original source - original target
2. original source - synthetic target (generated by 4 forward teachers)
3. synthetic source (generated by 4 backward teachers) - original target

Then, you can use all of those data to train a new teachers.

Does this make sense?

**Comment by emjotde** [comment]
I am closing this one as we have our own version of depth-scaling coming. Should arrive shortly.
